### PR TITLE
Add transfer vault capability for crab v2

### PR DESCRIPTION
**Pull request by KMKoushik**
# Task:
Add vault transferring capability to crabV2

I am getting `Contract code size exceeds 24576 bytes` so I disabled it for now and we have to fix it later

## Description

- [x] Add `transferVault` event
- [X] Set Cap to 0
- [X] Add tests for transfer vault
- [X] Add time lock contract
- [x] Add tests for timelock

### CODE REVIEW

**Review by aparnakr — APPROVED** [review]
Can you also add a simple test for this?

**Comment by aparnakr** [comment]
Do we want to add a timelock as well? If so you could look at https://github.com/compound-finance/compound-protocol/blob/master/contracts/Timelock.sol

**Comment by KMKoushik** [comment]
> Do we want to add a timelock as well? If so you could look at https://github.com/compound-finance/compound-protocol/blob/master/contracts/Timelock.sol

@aparnakr How bout making a separate PR for this?

File: packages/hardhat/contracts/strategy/CrabStrategyV2.sol [hunk 1]
```diff
@@ -193,15 +195,29 @@ contract CrabStrategyV2 is StrategyBase, StrategyFlashSwap, ReentrancyGuard, Own
...
+        emit VaultTransferred(_newStrategy, vaultId);
+    }
+
     /**
      * @notice owner can set the strategy cap in ETH collateral terms
      * @dev deposits are rejected if it would put the strategy above the cap amount
      * @dev strategy collateral can be above the cap amount due to hedging activities
      * @param _capAmount the maximum strategy collateral in ETH, checked on deposits
      */
     function setStrategyCap(uint256 _capAmount) external onlyOwner {
```
Comment (aleone) [inline packages/hardhat/contracts/strategy/CrabStrategyV2.sol:215]: Might be more simple to make this be a public function instead of having an external that calls internal function?  Get why you changed it, but public can have same result, probably with smaller bytecode and small gas differences (it may be slightly more expensive to call?)

I don't have strong opinions though, other thoughts from folks?  Its also possible we just keep in mind for future if we have bytecode size issues later on and can change it.

File: packages/hardhat/hardhat.config.ts [hunk 2]
```diff
@@ -74,6 +74,8 @@ const config: HardhatUserConfig = {
       accounts: {
         accountsBalance: '1000000000000000000000000000'
       },
+      // IMPORTANT: SHOULD BE REMOVED AND FIXED BEFORE MERGING
```
Comment (aleone) [inline packages/hardhat/hardhat.config.ts:77]: Easier way to fix this temp issue is to set the ```runs: 800``` under optimizer to 1 and see if that gets you under the limit and can raise it when we remove code.

**Review by aleone — APPROVED** [review]
Generally lgtm, few small comments. Ok doing the timelock in separate PR, but do think it may change things a decent amount / larger scope. Fine with either way of doing it.  

TLDR:
- Think easier to use optimizer to fix contract bytecode size for now, reverted later
- add missing natspec
- could consider public vs external/internal pattern especially if hit bytecode limit later

File: packages/hardhat/hardhat.config.ts (hunk 2 above)
Comment (aleone) [reply packages/hardhat/hardhat.config.ts:77]: (spoiler, it does get us under the limit :) )

File: packages/hardhat/contracts/strategy/CrabStrategyV2.sol (hunk 1 above)
Comment (KMKoushik) [reply packages/hardhat/contracts/strategy/CrabStrategyV2.sol:215]: The only reason I used `external` is that I read it's cheap. But yeah it's not frequently called function and doesn't necessarily have to be external. Will change!

File: packages/hardhat/hardhat.config.ts (hunk 2 above)
Comment (KMKoushik) [reply packages/hardhat/hardhat.config.ts:77]: Yep worked

**Comment by KMKoushik** [comment]
@aparnakr @aleone Lot's of changes needed in vault transfer logic to implement timelock. So going to use the same PR. Not ready yet

**Comment by aparnakr** [comment]
> > Do we want to add a timelock as well? If so you could look at https://github.com/compound-finance/compound-protocol/blob/master/contracts/Timelock.sol
> 
> @aparnakr How bout making a separate PR for this?

yes makes sense

File: packages/hardhat/contracts/strategy/timelock/SafeMath.sol [hunk 3]
```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: BSD-3-Clause
```
Comment (KMKoushik) [inline packages/hardhat/contracts/strategy/timelock/SafeMath.sol:1]: Don't want to change any code from the compound, so I copied math lib too

File: packages/hardhat/hardhat.config.ts [hunk 4]
```diff
@@ -212,7 +212,18 @@ const config: HardhatUserConfig = {
   solidity: {
     compilers: [
       UNISWAP_SETTING,
-    ]
+    ],
+    overrides: {
```
Comment (KMKoushik) [inline packages/hardhat/hardhat.config.ts:216]: Timelock related contracts need `0.8.10` as it's used by compound

File: packages/hardhat/test/integration-tests/crabv2/crab-timelock.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+      currentBlockNumber = await provider.getBlockNumber()
+      const currentBlock = await provider.getBlock(currentBlockNumber)
+      eta = currentBlock.timestamp + twoDays + 300
+      currentBlockTimestamp = currentBlock.timestamp
+
+      data = abi.encode(['address'], [depositor2.address]);
+
+      txHash = ethers.utils.keccak256(abi.encode(['address', 'uint256', 'string', 'bytes', 'uint256'],
+        [crabStrategy.address, '0', signature, data, eta]))
+      await timelock.connect(owner).queueTransaction(crabStrategy.address, 0, signature, data, eta)
```
Comment (aparnakr) [inline packages/hardhat/test/integration-tests/crabv2/crab-timelock.ts:187]: could you make this a test instead of in the before all and check that it doesn't work if we accidentally pass in bad data?

**Review by aparnakr — APPROVED** [review]
Are the tests also mostly from Compound? It might be fine to just port over and edit their tests.

**Review by aleone — APPROVED** [review]
lgtm

**Comment by KMKoushik** [comment]
> Are the tests also mostly from Compound? It might be fine to just port over and edit their tests. 

That's what I did almost

**Comment by KMKoushik** [comment]
> Is the contract size below the limit now? 

Yep

**Comment by KMKoushik** [comment]
Will be merged in #530
